browser/brave_ads/creatives/search_result_ad/creative_search_result_ad_tab_helper.cc

```diff
@@ -0,0 +1,263 @@
+  CHECK(web_contents);
+
+  if (!web_contents->GetBrowserContext() ||
+      web_contents->GetBrowserContext()->IsOffTheRecord()) {
+    return;
+  }
+
+  if (!ShouldSupportSearchResultAds()) {
+    return;
+  }
+
+  CreateForWebContents(web_contents);
+}
+
+bool CreativeSearchResultAdTabHelper::ShouldHandleCreativeAdEvents() const {
+  if (!ShouldSupportSearchResultAds()) {
+    // If the feature is disabled, we should not trigger creative ad events.
+    return false;
+  }
+
+  if (ShouldAlwaysTriggerSearchResultAdEvents()) {
+    // If the feature is enabled, we should always trigger creative ad events.
+    return true;
+  }
+
+  // If the feature is enabled, we should only trigger creative ad events when
+  // the user has joined Brave Rewards.
+  const Profile* const profile =
+      Profile::FromBrowserContext(web_contents()->GetBrowserContext());
+  return profile->GetPrefs()->GetBoolean(brave_rewards::prefs::kEnabled);
+}
+
+void CreativeSearchResultAdTabHelper::MaybeTriggerCreativeAdClickedEvent(
+    const GURL& url) {
+  if (creative_search_result_ad_handler_) {
+    creative_search_result_ad_handler_->MaybeTriggerCreativeAdClickedEvent(url);
+  }
+}
+
+///////////////////////////////////////////////////////////////////////////////
+
+AdsService* CreativeSearchResultAdTabHelper::GetAdsService() const {
+  if (g_ads_service_for_testing) {
+    CHECK_IS_TEST();
+
+    return g_ads_service_for_testing;
+  }
+
+  Profile* const profile =
+      Profile::FromBrowserContext(web_contents()->GetBrowserContext());
+  return AdsServiceFactory::GetForProfile(profile);
+}
+
+void CreativeSearchResultAdTabHelper::MaybeCreateCreativeSearchResultAdHandler(
+    content::NavigationHandle* navigation_handle) {
+  CHECK(navigation_handle);
+
+  if (!ShouldHandleCreativeAdEvents()) {
+    return;
+  }
+
+  // Do not trigger ad viewed events if the user navigates back or forward.
+  const bool should_trigger_creative_ad_viewed_events =
+      (navigation_handle->GetPageTransition() &
+       ui::PAGE_TRANSITION_FORWARD_BACK) == 0;
+
+  creative_search_result_ad_handler_ =
+      CreativeSearchResultAdHandler::MaybeCreate(
+          GetAdsService(), navigation_handle->GetURL(),
+          should_trigger_creative_ad_viewed_events);
+}
+
+void CreativeSearchResultAdTabHelper::
+    MaybeExtractCreativeAdPlacementIdsFromWebPageAndHandleViewedEvents() {
+  if (!ShouldHandleCreativeAdEvents()) {
+    return;
+  }
+
+  if (!creative_search_result_ad_handler_) {
+    return;
+  }
+
+  creative_search_result_ad_handler_
+      ->MaybeExtractCreativeAdPlacementIdsFromWebPage(
+          web_contents()->GetPrimaryMainFrame(),
+          base::BindOnce(&CreativeSearchResultAdTabHelper::
+                             MaybeHandleCreativeAdViewedEvents,
+                         weak_factory_.GetWeakPtr()));
+}
+
+void CreativeSearchResultAdTabHelper::MaybeHandleCreativeAdViewedEvents(
+    std::vector<std::string> placement_ids) {
```

Comment:
const can be added: `const std::vector<std::string> placement_ids`

Reply:
Fixed